Coming Soon: Render a coming soon page if the user is logged out

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

* Gate feature behind a `WPCOM_PUBLIC_COMING_SOON` config flag
* Render coming soon page if `wpcom_public_coming_soon` is `1` and user isn't a logged in site user
* Render `wpcom_public_coming_soon_page_id` if it exists
* Render a fallback coming soon page if `wpcom_public_coming_soon_page_id` doesn't exist

Based off @simison's prototype #43831

What it doesn't do:
* check if the page is published
* check if it's a post vs. a page, maybe this doesn't matter?

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout branch and `cd apps/editing-toolkit && yarn dev`
* Add `"WPCOM_PUBLIC_COMING_SOON": "1"` to your `.wp-env.override.json`
* `npx wp-env start`
* Visit `http://localhost:4013/wp-admin` and create a coming soon page in the block editor
* `npx wp-env run cli wp option add wpcom_public_coming_soon 1`
* Visit `http://localhost:4013` in an incognito window, it should always show a generic coming soon page
* `npx wp-env run cli wp option add wpcom_public_coming_soon_page_id {{ YOUR PAGE ID }}`
* Visit `http://localhost:4013` in an incognito window, it should render a page that's generally like the page you created (it doesn't show some menus or footers, but it should look generally correct)

andrewserong commented:
@p-jackson I've rebased this one and made the following changes:

* Copied over the fallback coming soon page from #45167 
* Renamed the option to `wpcom_coming_soon_page_id` to match D45700-code (happy to switch this back to `wpcom_coming_soon_page` if you prefer that one instead!
* Added check for `wpcom_coming_soon` option in `show_coming_soon_page`
* Added `current_user_can( 'read' )` capability check, which should deal with the case of someone who's logged in but aren't a member of the site

Another thing I liked that @ramonjd added in #45167 was grabbing the post content for the coming soon page and injecting it into an HTML shell, instead of doing a redirect so that we don't display the header and footer of the site in the custom coming soon page. I haven't copied that in here, because I thought you might have preferred the redirect approach and there might be another way of going about this? Happy to copy that over too, if you want to go in that direction.

andrewserong commented on apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php:
```diff
@@ -56,22 +56,44 @@ function coming_soon_page() {
+		</head>
+		<body>
+			<?php
+				// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Is it okay to silence this line? Because we're just echoing out the post_content applied through `the_content`, is that safe enough?

p-jackson replied:
I copied what the `the_content()` function does. Doesn't fix the warning unfortunately, but I couldn't find a core escaping function that did the same thing as `str_replace( ']]>', ']]&gt;', ... )`

andrewserong commented:
Just updated again with swapping out the redirect for rendering the post content in an HTML shell.

* [ ] Still need to suppress the share buttons from being shown, too.

andrewserong commented:
Just adding some additional testing instructions. For testing in my sandbox, I've been using the following in wpsh:

```
$test_site_id = wpcom_get_blog_id( 'andysfakecomingsoonv2testsite.wordpress.com' );
update_blog_option( $test_site_id, 'wpcom_coming_soon', 1 );
update_blog_option( $test_site_id, 'wpcom_coming_soon_page_id', 36 ); // the id of my custom coming soon page

// To switch off the custom page and view the fallback:
update_blog_option( $test_site_id, 'wpcom_coming_soon_page_id', 0 ); // the id of my custom coming soon page
```

p-jackson commented:
> * [ ] Still need to suppress the share buttons from being shown, too.

I reckon we should merge and finesse as we go

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php:
```diff
@@ -315,3 +315,11 @@ function load_wpcom_block_editor_sidebar() {
+/**
+ * Coming soon
+ */
+function load_coming_soon() {
```
Will we need any feature flag here to prevent the module loading? Or are we just relying on the non-existence of the coming soon options.

p-jackson commented:
I _think_ we can just rely on the site option not existing 🤞😕🤞
What do you think @andrewserong? Does it seem safe enough?

andrewserong commented:
@p-jackson just looking over this again, and I think we might want to set a feature flag of some kind, because in https://github.com/Automattic/wp-calypso/pull/44888/files#diff-1cf0a20f37e4c4e09a0f16925cff4f6fR23 it looks like this will override existing sites set to Coming Soon by sending them to the fallback mode. We should be able to test that though, grab an existing site set to v1 Coming Soon, and see if this line appears on the front end: https://github.com/Automattic/wp-calypso/pull/44888/files#diff-331b8751885bccfd046f3c082e31fa2eR191

But while we're still developing the feature, it just might feel safer for us if we have an explicit check, just in case?

p-jackson commented:
> it looks like this will override existing sites set to Coming Soon by sending them to the fallback mode

Oh thank you for reminding me! I had forgotten that this was the same flag that's being used by v1! I'm inclined to just use a completely different flag for v2. `wpcom_coming_soon_v2` or someting.

andrewserong commented:
I'd been wondering about maybe `wpcom_public_coming_soon` or `wpcom_coming_soon_public`? I was a little torn over whether or not there's a benefit to re-using the existing site option (would there be more stuff we need to update or replace if we use a new site option?), but it would certainly be a neat way to gate the feature keeping it separate.

p-jackson commented:
> I was a little torn over whether or not there's a benefit to re-using the existing site option

I'm definitely leaning towards treating them like completely different features rather than two versions of the same feature.

> I'd been wondering about maybe `wpcom_public_coming_soon` or `wpcom_coming_soon_public`?

Yes! I like putting public in the name. "public coming soon" is probably a better project label than "v2" as well :D

p-jackson commented:
I've renamed the site options to `wpcom_public_coming_soon` and `wpcom_public_coming_soon_page_id`. I've also added a feature gate, so the code doesn't get loaded unless the `WPCOM_PUBLIC_COMING_SOON` config flag is set. So we can be more confident that we won't break anything, but I think once we're confident enough to start testing on atomic we'll probably want to just remove the config flag I think, because it's fiddly setting a config flag on atomic.

You can set the config either in `.wp-env.override.json` or `0-sandbox.php` depending on where you're testing.

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/coming-soon/fallback-coming-soon-page.php:
```diff
@@ -0,0 +1,208 @@
+				<?php if ( ! is_user_logged_in() ) : ?>
+					<div class="marketing-copy">
+						<img src="/wp-content/themes/a8c/domain-landing-page/wpcom-wmark-white.svg" alt="WordPress.com" class="logo" />
+						<p class="copy"><?php echo esc_html( fix_widows( __( 'Build a website. Sell your stuff. Write a blog. And so much more.', 'full-site-editing' ), array( 'mobile_enable' => true ) ) ); ?></p>
```
Not sure if this is in the scope of this PR, but will assumed WordPress.com globals such as these work in Atomic? 

`fix_widows` exists on WordPress.com as a blog-plugin but not is undefined in wp-env 

<img width="1045" alt="Screen Shot 2020-09-09 at 1 23 37 pm" src="https://user-images.githubusercontent.com/6458278/92550777-afb7c680-f29f-11ea-87fd-f4f25ae943e7.png">

Same for `$login_url` and `localized_wpcom_url()` below.

p-jackson replied:
Maybe we should wait to see wether we're going to redesign the fallback coming soon page before worrying about fixing widows.

Missing `$login_url` and `localized_wpcom_url()` is interesting! I wonder if we should just the hide the banner if they're missing? Maybe another design/product question. It's kind of like the "Powered by WordPress.com" links not being on higher plans. Maybe if they're on Atomic then we can hide the WordPress.com banner because they have a plan?

simison replied:
Is it risk free now to deploy this at Atomic? I.e. this code won't be run?

p-jackson replied:
It can't run unless `WPCOM_PUBLIC_COMING_SOON` is defined, so it'll only run if devs manually set that constant on their Atomic site to try it out.

ramonjd reviewed (approved):
Basic functionality LGTM

Tested with feature flag on and off in wp-override.json

Just one question about pulling over the globals available in WordPress.com but not elsewhere.

simison commented on apps/editing-toolkit/editing-toolkit-plugin/coming-soon/fallback-coming-soon-page.php:
```diff
@@ -0,0 +1,208 @@
+					padding-bottom: 32px;
+				}
+			}
+		</style>
```
Since this is quite a bit of inline CSS, it might make sense to have it as a separate file and put through build-pipeline. Then either enqueue it as a file or inject as inline CSS.
 
Benefits:
- You get linting
- You can use sass, which would be important for using variables to keep things more consistent. 
- You get Autoprefixer so no human errors with vendor specific CSS
- You get minification (could be we do that for inline CSS but I haven't checked)

andrewserong replied:
Great idea! For the moment, I just copied over the existing coming soon page verbatim from wpcom, but it'd be good to fix all this up. There's some styling overrides we'll likely want to add too, to make sure that theme styling doesn't mess with the login buttons on the mobile view, too.
